File: Data Structures & Algorithms/combination-sum-iv/submission-1.cpp
class Solution {
public:
    int combinationSum4(vector<int>& nums, int target) {
        // dp[i][j]: combinations for using numbers from s[:i] sum up to j

        vector<int> dp(target+1, 0);
        // base case
        dp[0] = 1;

        for (int total = 1; total <= target; total++) {
            for (int num:nums) {
                if (total >= num) {
                    dp[total] += dp[total - num];
                }
            }
        }
        return dp[target];
    }
};

File: Data Structures & Algorithms/split-array-largest-sum/submission-1.py
class Solution:
    def splitArray(self, nums: List[int], k: int) -> int:
            
        def canSplit(target) -> bool:
            currentSum = 0
            splits = 1
            for num in nums:
                currentSum += num
                if currentSum > target:
                    splits += 1
                    currentSum = num
                    if splits > k:
                        return False
            return True
        
        l = max(nums)
        r = sum(nums)
        ans = 0
        while l <= r:
            mid = (l + r)//2

            if canSplit(mid):
                ans = mid
                r = mid - 1
            else:
                l = mid + 1

        return ans

        